### PR TITLE
add FastLogger and related helpers

### PR DESCRIPTION
**Pull request by winterland1989**
This is the first step to solve #84, The higher level api then can be used to add `runFastLoggerLoggingT` to monad-logger.

One thing i'm not sure is that i have copied Date.hs from wai-logger to get cached formatted time, i'm not sure i should do this or update it in date-cache and add it as a dependency.

### CODE REVIEW

**Comment by winterland1989** [comment]
@kazu-yamamoto Please review : )

**Comment by winterland1989** [comment]
Sorry for the commit noise, i didn't expect wai-logger will break if i export `LogType` from fast-logger, which i should. Now everything should be fine. If these change are ok to you, i will process send pr to monad-logger, then we can bump version together.

File: fast-logger/System/Log/FastLogger.hs [hunk 1]
```diff
@@ -1,7 +1,7 @@
 -- | This module provides a fast logging system which
 --   scales on multicore environments (i.e. +RTS -N\<x\>).
 {-# LANGUAGE CPP #-}
-{-# LANGUAGE Safe #-}
```
Comment (kazu-yamamoto) [inline fast-logger/System/Log/FastLogger.hs]: Why did you delete `Safe`?

Comment (winterland1989) [inline fast-logger/System/Log/FastLogger.hs]: Because GHC tell me i can't use unix package while mark module safe, because unix is not safe.

File: fast-logger/System/Log/FastLogger.hs [hunk 2]
```diff
@@ -150,3 +164,125 @@ rmLoggerSet (LoggerSet mfile fref arr _) = do
...
+                when (cnt <= 0) $ tryRotate lgrset fspec ref mvar
+        return (logger, rmLoggerSet lgrset)
+
+-- | 'bracket' version of 'newFastLogger'
+withFastLogger :: LogType -> (FastLogger -> IO a) -> IO ()
+withFastLogger typ log' = bracket (newFastLogger typ) (log' . fst) snd
+
+-- | Initialize a 'FastLogger' with timestamp attached to each message.
+-- a tuple of logger and clean up action are returned.
+newTimedFastLogger :: TimeFormat -> LogType -> IO (TimedFastLogger, IO ())
```
Comment (winterland1989) [inline fast-logger/System/Log/FastLogger.hs]: Do you think i should change `TimeFormat` to `IO FormattedTime` here? so that you can share formatted time cache across logger.

Comment (winterland1989) [inline fast-logger/System/Log/FastLogger.hs]: Do you think i should change `TimeFormat` to `IO FormattedTime` here? so that you can share formatted time cache across logger.

File: fast-logger/System/Log/FastLogger.hs [hunk 3]
```diff
@@ -150,3 +164,125 @@ rmLoggerSet (LoggerSet mfile fref arr _) = do
...
+                when (cnt <= 0) $ tryRotate lgrset fspec ref mvar
+        return (logger, rmLoggerSet lgrset)
+
+-- | 'bracket' version of 'newTimeFastLogger'
+withTimedFastLogger ::  TimeFormat -> LogType -> (TimedFastLogger -> IO a) -> IO ()
+withTimedFastLogger fmt typ log' = bracket (newTimedFastLogger fmt typ) (log' . fst) snd
+
+-- | A simple time format: @simpleTimeFormat = "%d/%b/%Y:%T %z"@
+simpleTimeFormat :: TimeFormat
+simpleTimeFormat = "%d/%b/%Y:%T %z"
```
Comment (winterland1989) [inline fast-logger/System/Log/FastLogger.hs]: I'd like to move `simpleTimeFormat` to date module, should we?

File: fast-logger/System/Log/FastLogger.hs (hunk 1 above)
Comment (kazu-yamamoto) [inline fast-logger/System/Log/FastLogger.hs]: OK. Thank you for your explanation.

**Comment by kazu-yamamoto** [comment]
The quality seems good to me. So, I have merged.

File: wai-logger/Network/Wai/Logger.hs [hunk 4]
```diff
@@ -96,26 +96,17 @@ data ApacheLoggerActions = ApacheLoggerActions {
...
-             | LogStdout BufSize           -- ^ Logging to stdout.
-                                           --   'BufSize' is a buffer size
-                                           --   for each capability.
-             | LogFile FileLogSpec BufSize -- ^ Logging to a file.
-                                           --   'BufSize' is a buffer size
-                                           --   for each capability.
-                                           --   File rotation is done on-demand.
-             | LogCallback (LogStr -> IO ()) (IO ())
-
 ----------------------------------------------------------------
```
Comment (winterland1989) [inline wai-logger/Network/Wai/Logger.hs]: Actually i'd like to refract wai-logger completely, remove Date.hs and use new fast-logger implementation, export `clockDateCacher = newTimeCacher simpleTimeFormat`, but i'm not sure if i can safely remove `ZonedDate/DateCacheGetter/DateCacheUpdater`. AFAICT no package use these exports.

File: fast-logger/System/Log/FastLogger.hs (hunk 3 above)
Comment (kazu-yamamoto) [inline fast-logger/System/Log/FastLogger.hs]: Probably, yes. :-)

File: wai-logger/Network/Wai/Logger.hs (hunk 4 above)
Comment (kazu-yamamoto) [inline wai-logger/Network/Wai/Logger.hs]: Please just do it. :-)

File: fast-logger/System/Log/FastLogger.hs (hunk 2 above)
Comment (kazu-yamamoto) [inline fast-logger/System/Log/FastLogger.hs]: Would you show me code to demonstrate the share with `IO FormattedTime`?

**Comment by kazu-yamamoto** [comment]
After releasing fast-logger 2.4.4, I noticed that this version breaks backward compatibility from the wai-logger point of view. The signature of `LogFile` is changed. I would suggest:

s/LogFile/LogFileNoRotate/
s/LogFileAutoRotate/LogFile/

**Comment by kazu-yamamoto** [comment]
I  made fast-logger-2.4.4 deprecated.

**Comment by winterland1989** [comment]
I'm just about to tell you this, and the `with...` version misuse `bracket` also, i'm working on a new pr.

**Comment by kazu-yamamoto** [comment]
Thanks.

File: wai-logger/Network/Wai/Logger.hs (hunk 4 above)
Comment (felixonmars) [inline wai-logger/Network/Wai/Logger.hs]: Turns out yesod-core is using DateCacheGetter: https://github.com/yesodweb/yesod/issues/1214

Comment (winterland1989) [inline wai-logger/Network/Wai/Logger.hs]: Oh，that is too bad， should we fix it here or change yesod cabal？
